apps/plugins/zfs/plugin/flist.go

```diff
@@ -0,0 +1,265 @@
+package main
+
+import (
+	"archive/tar"
+	"compress/bzip2"
+	"compress/gzip"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"syscall"
+)
+
+const (
+	//HubHost host name of hub
+	HubHost = "hub.grid.tf"
+
+	//ValidFileName
+	ValidFileName = ".valid"
+)
+
+type FList interface {
+	Hash() (string, error)
+	Open() (io.ReadCloser, error)
+	Ext() string
+}
+
+//LocalFList represents a local flist
+type LocalFList string
+
+func (l LocalFList) Hash() (string, error) {
+	_, err := os.Stat(string(l))
+	if err != nil {
+		return "", err
+	}
+
+	m := md5.New()
+	io.WriteString(m, string(l))
+	return fmt.Sprintf("%x", m.Sum(nil)), nil
+}
+
+func (l LocalFList) Open() (io.ReadCloser, error) {
+	return os.Open(string(l))
+}
+
+func (l LocalFList) Ext() string {
+	return path.Ext(string(l))
+}
+
+type RemoteFList string
+
+func (l RemoteFList) Hash() (string, error) {
+	resp, err := http.Get(fmt.Sprintf("%s.md5", string(l)))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	data, err := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to get flist hash: %s", resp.Status)
+	}
+
+	return string(data[:32]), nil
+}
+
+func (l RemoteFList) Open() (io.ReadCloser, error) {
+	response, err := http.Get(string(l))
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to download flist: %s", response.Status)
+	}
+
+	return response.Body, nil
+}
+
+func (l RemoteFList) Ext() string {
+	return path.Ext(string(l))
+}
+
+type Meta struct {
+	Base string
+	Hash string
+}
+
+func getFList(src string) (FList, error) {
+	u, err := url.Parse(src)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.Scheme == "file" || u.Scheme == "" {
+		return LocalFList(u.Path), nil
+	} else if u.Scheme == "http" || u.Scheme == "https" {
+		if u.Hostname() != HubHost {
```

Comment:
Why do we enforce this ?
Notice prevent the user to host their flist somewhere else.

Reply:
I want to grantee that a `'%s.md5' % flist_url` always exists. so only our hub has this. If u try now any flist from a different location (any http server) it probably won't have this unless it's running an instance of the hub.

Anyway, this limitation is more of a sanity check, that can be loosen later on